### ISSUE TITLE
Query regarding Extension in Server Hello Message TLS v1.2

### ISSUE THREAD

**Issue by csanoop** [openssl/openssl#3530]
Hi ,
This is a query regarding the behavior of TLSv1.2 client in case Server Hello message contains additional extension not requested by client. Detailed scenario is furnished below and any lead would be appreciated,

From the RFC5246 [TLS v12], I see the below.

> _extensions
> A list of extensions. Note that only extensions offered by the client can appear in the server’s list._

Based on the above statement, I interpreted that server sends the server hello message with extensions which were only offered in Client hello message by Client.

In case of openssl server (version -1.1.0d), the server hello message contains the extensions as specified below [Server Hello] for the respective client hello message with the extensions specified below [Client Hello]. 

Server Hello Msg: Extension information from server

> 00 08 
>  00 0b 00 04 03 **_00 01 02_**

Client Hello Msg: Extension information [Client supports only uncompressed point format]

> 00 16 
> 00 0a 00 04 00 02 00 17
> 00 0d 00 04 00 02 04 03
> 00 0b 00 02 01 **_00_**

Here the server hello message contains the extensions [_**01 02**_] which are not specified in the client hello message.
So considering the above statements from the RFC, what is the expected behavior of the client in this case when it observes unsupported extensions in the server hello message?

**Comment by mattcaswell** [comment]
The extension you are talking about here is the "Supported Point Formats" extension as defined in RFC4492. The *contents* of that extension is a list of supported point formats, i.e. one of uncompressed, ansiX962_compressed_prime and ansiX962_compressed_char2.

The wording you quote from RFC5246 refers to the presence or absence of the whole extension, i.e. if the client does not send the "supported point formats" extension then the server isn't supposed to send one back. It does not refer to the contents of that extension. What is sent back in the contents is defined individually by each extension. In this case RFC4492 says this:

    The server's Supported Point Formats Extension has the same structure
    as the client's Supported Point Formats Extension (see
    Section 5.1.2).  Items in elliptic_curve_list here are ordered
    according to the server's preference (favorite choice first).  Note
    that the server may include items that were not found in the client's
    list (e.g., the server may prefer to receive points in compressed
    format even when a client cannot parse this format: the same client
    may nevertheless be capable of outputting points in compressed
    format).

So in other words the client says what types of points it is capable of parsing, and the server also says what types of points it is capable of parsing. The two lists may be different, which is fine. The extension says nothing about what the client/server are capable of *sending*.
